File: scripts/parity/query/node/dump.test.ts
import { describe, it, expect } from "vitest";
import { spawnSync } from "node:child_process";
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import Ajv from "ajv/dist/2020.js";

// Integration tests for dump.ts — runs the script against real fixtures.

const SCRIPT_DIR = dirname(fileURLToPath(import.meta.url));
const DUMP_SCRIPT = join(SCRIPT_DIR, "dump.ts");
const FIXTURES = resolve(SCRIPT_DIR, "../../fixtures");
const TSX_BIN = resolve(
  SCRIPT_DIR,
  "../../../../node_modules/.bin",
  process.platform === "win32" ? "tsx.cmd" : "tsx",
);

interface RunResult {
  code: number;
  stdout: string;
  stderr: string;
  outPath: string;
}

function runDump(fixture: string, opts: { frozenAt?: string } = {}): RunResult {
  const outDir = mkdtempSync(join(tmpdir(), "parity-node-test-"));
  const outPath = join(outDir, `${fixture}.json`);
  const args = [DUMP_SCRIPT, join(FIXTURES, fixture), outPath];
  if (opts.frozenAt) args.push("--frozen-at", opts.frozenAt);
  const res = spawnSync(TSX_BIN, args, { encoding: "utf8" });
  return {
    code: res.status ?? -1,
    stdout: res.stdout ?? "",
    stderr: res.stderr ?? "",
    outPath,
  };
}

function cleanup(outPath: string): void {
  try {
    rmSync(dirname(outPath), { recursive: true, force: true });
  } catch {
    /* ignore */
  }
}

const DEFAULT_FROZEN_AT = "2000-01-01T00:00:00.000Z";

describe("dump.ts", () => {
  it("dumps arel-01 (Table) with expected CanonicalQuery fields", () => {
    const { code, stdout, stderr, outPath } = runDump("arel-01");
    try {
      expect(code, `dump failed\nstdout: ${stdout}\nstderr: ${stderr}`).toBe(0);
      const result = JSON.parse(readFileSync(outPath, "utf8"));
      expect(result.version).toBe(1);
      expect(result.fixture).toBe("arel-01");
      expect(result.frozenAt).toBe(DEFAULT_FROZEN_AT);
      expect(result.sql).toMatch(/"users"/i);
      expect(result.binds).toEqual([]);
    } finally {
      cleanup(outPath);
    }
  });

  it("dumps arel-06 (eq predicate)", () => {
    const { code, stdout, stderr, outPath } = runDump("arel-06");
    try {
      expect(code, `dump failed\nstdout: ${stdout}\nstderr: ${stderr}`).toBe(0);
      const result = JSON.parse(readFileSync(outPath, "utf8"));
      expect(result.sql).toMatch(/"users"\."name" = /i);
    } finally {
      cleanup(outPath);
    }
  });

  it("dumps arel-09 (lt predicate)", () => {
    const { code, stdout, stderr, outPath } = runDump("arel-09");
    try {
      expect(code, `dump failed\nstdout: ${stdout}\nstderr: ${stderr}`).toBe(0);
      const result = JSON.parse(readFileSync(outPath, "utf8"));
      expect(result.sql).toMatch(/"users"\."age" < /i);
    } finally {
      cleanup(outPath);
    }
  });

  it("dumps arel-21 (SelectManager with WHERE)", () => {
    const { code, stdout, stderr, outPath } = runDump("arel-21");
    try {
      expect(code, `dump failed\nstdout: ${stdout}\nstderr: ${stderr}`).toBe(0);
      const result = JSON.parse(readFileSync(outPath, "utf8"));
      expect(result.sql).toMatch(/SELECT/i);
      expect(result.sql).toMatch(/WHERE/i);
    } finally {
      cleanup(outPath);
    }
  });

  it("preserves --frozen-at string verbatim in frozenAt output", () => {
    const frozen = "2026-01-01T00:00:00.000Z";
    const { code, stdout, stderr, outPath } = runDump("arel-01", { frozenAt: frozen });
    try {
      expect(code, `dump failed\nstdout: ${stdout}\nstderr: ${stderr}`).toBe(0);
      const result = JSON.parse(readFileSync(outPath, "utf8"));
      expect(result.frozenAt).toBe(frozen);
    } finally {
      cleanup(outPath);
    }
  });

  it("exits 1 with stderr when --frozen-at has no value", () => {
    const outDir = mkdtempSync(join(tmpdir(), "parity-node-test-"));
    const outPath = join(outDir, "out.json");
    try {
      const res = spawnSync(
        TSX_BIN,
        [DUMP_SCRIPT, join(FIXTURES, "arel-01"), outPath, "--frozen-at"],
        { encoding: "utf8" },
      );
      expect(res.status).toBe(1);
      expect(res.stderr).toMatch(/--frozen-at requires a value/);
    } finally {
      rmSync(outDir, { recursive: true, force: true });
    }
  });

  it("exits 1 with stderr when --frozen-at has invalid format", () => {
    const { code, stderr, outPath } = runDump("arel-01", { frozenAt: "not-a-timestamp" });
    try {
      expect(code).toBe(1);
      expect(stderr).toMatch(/--frozen-at must be ISO 8601 UTC with trailing Z/);
    } finally {
      cleanup(outPath);
    }
  });

  it("emits output that validates against query.schema.json", () => {
    const schemaPath = resolve(SCRIPT_DIR, "../../canonical/query.schema.json");
    const schema = JSON.parse(readFileSync(schemaPath, "utf8"));
    const ajv = new Ajv({ strict: false });
    const validate = ajv.compile(schema);

    const { code, stdout, stderr, outPath } = runDump("arel-06");
    try {
      expect(code, `dump failed\nstdout: ${stdout}\nstderr: ${stderr}`).toBe(0);
      const result = JSON.parse(readFileSync(outPath, "utf8"));
      const ok = validate(result);
      expect(ok, `schema validation failed: ${JSON.stringify(validate.errors, null, 2)}`).toBe(
        true,
      );
    } finally {
      cleanup(outPath);
    }
  });
});

File: scripts/parity/query/ruby/dump_test.rb
# frozen_string_literal: true

require "bundler/setup"
require "minitest/autorun"
require "json"
require "tempfile"
require "open3"

# Integration tests for dump.rb — runs the script against real fixtures.
# Must be run under Bundler with scripts/parity/schema/ruby/Gemfile.

GEMFILE     = File.expand_path("../../schema/ruby/Gemfile", __dir__)
DUMP_SCRIPT = File.expand_path("dump.rb", __dir__)
FIXTURES    = File.expand_path("../../fixtures", __dir__)

def run_dump(fixture, frozen_at: nil)
  tf = Tempfile.new(["parity-test-", ".json"])
  out = tf.path
  tf.close  # close so ruby on the other side can write to it
  cmd = ["bundle", "exec", "--gemfile=#{GEMFILE}", "ruby", DUMP_SCRIPT,
         "#{FIXTURES}/#{fixture}", out]
  cmd += ["--frozen-at", frozen_at] if frozen_at
  stdout, stderr, status = Open3.capture3(*cmd)
  [status.exitstatus, stdout, stderr, out]
end

class DumpTest < Minitest::Test
  DEFAULT_FROZEN_AT = "2000-01-01T00:00:00.000Z"

  def test_arel_01_table_object
    code, stdout, stderr, out_path = run_dump("arel-01")
    assert_equal 0, code, "dump failed\nstdout: #{stdout}\nstderr: #{stderr}"
    result = JSON.parse(File.read(out_path))
    assert_equal 1,                result["version"]
    assert_equal "arel-01",        result["fixture"]
    assert_equal DEFAULT_FROZEN_AT, result["frozenAt"]
    assert_match(/"users"/i,       result["sql"])
    assert_equal [],               result["binds"]
  ensure
    File.delete(out_path) if File.exist?(out_path)
  end

  def test_arel_06_eq_predicate
    code, stdout, stderr, out_path = run_dump("arel-06")
    assert_equal 0, code, "dump failed\nstdout: #{stdout}\nstderr: #{stderr}"
    result = JSON.parse(File.read(out_path))
    assert_match(/"users"\."name" = /i, result["sql"])
  ensure
    File.delete(out_path) if File.exist?(out_path)
  end

  def test_arel_09_lt_predicate
    code, stdout, stderr, out_path = run_dump("arel-09")
    assert_equal 0, code, "dump failed\nstdout: #{stdout}\nstderr: #{stderr}"
    result = JSON.parse(File.read(out_path))
    assert_match(/"users"\."age" < /i, result["sql"])
  ensure
    File.delete(out_path) if File.exist?(out_path)
  end

  def test_arel_21_select_manager_with_where
    # arel-21 returns a SelectManager — exercises the to_sql_and_binds branch
    code, stdout, stderr, out_path = run_dump("arel-21")
    assert_equal 0, code, "dump failed\nstdout: #{stdout}\nstderr: #{stderr}"
    result = JSON.parse(File.read(out_path))
    assert_match(/SELECT/i, result["sql"], "expected SELECT statement from SelectManager")
    assert_match(/WHERE/i,  result["sql"], "expected WHERE clause")
  ensure
    File.delete(out_path) if File.exist?(out_path)
  end

  def test_frozen_at_forwarded
    frozen = "2026-01-01T00:00:00.000Z"
    code, stdout, stderr, out_path = run_dump("arel-01", frozen_at: frozen)
    assert_equal 0, code, "dump failed\nstdout: #{stdout}\nstderr: #{stderr}"
    result = JSON.parse(File.read(out_path))
    assert_equal frozen, result["frozenAt"]
  ensure
    File.delete(out_path) if File.exist?(out_path)
  end

  def test_frozen_at_missing_value
    tf = Tempfile.new(["parity-test-", ".json"])
    out = tf.path
    tf.close
    cmd = ["bundle", "exec", "--gemfile=#{GEMFILE}", "ruby", DUMP_SCRIPT,
           "#{FIXTURES}/arel-01", out, "--frozen-at"]
    _stdout, stderr, status = Open3.capture3(*cmd)
    assert_equal 1, status.exitstatus
    assert_match(/--frozen-at requires a value/, stderr)
  ensure
    File.delete(out) if File.exist?(out)
  end

  def test_frozen_at_invalid_format
    tf = Tempfile.new(["parity-test-", ".json"])
    out = tf.path
    tf.close
    cmd = ["bundle", "exec", "--gemfile=#{GEMFILE}", "ruby", DUMP_SCRIPT,
           "#{FIXTURES}/arel-01", out, "--frozen-at", "not-a-timestamp"]
    _stdout, stderr, status = Open3.capture3(*cmd)
    assert_equal 1, status.exitstatus
    assert_match(/--frozen-at must be ISO 8601 UTC with trailing Z/, stderr)
  ensure
    File.delete(out) if File.exist?(out)
  end
end
